File: src/lib/java/work/lclpnet/ap2/impl/game/kit/KitHandler.java
package work.lclpnet.ap2.impl.game.kit;

import com.mojang.serialization.Codec;
import com.mojang.serialization.MapCodec;
import lombok.Getter;
import net.minecraft.ChatFormatting;
import net.minecraft.core.component.DataComponents;
import net.minecraft.server.level.ServerLevel;
import net.minecraft.server.level.ServerPlayer;
import net.minecraft.sounds.SoundEvents;
import net.minecraft.sounds.SoundSource;
import net.minecraft.world.InteractionResult;
import net.minecraft.world.item.Item;
import net.minecraft.world.item.ItemStack;
import net.minecraft.world.item.Items;
import work.lclpnet.ap2.api.base.Participants;
import work.lclpnet.ap2.api.game.MiniGameHandle;
import work.lclpnet.ap2.impl.game.GameCommons;
import work.lclpnet.ap2.impl.util.CustomNbt;
import work.lclpnet.kibu.access.entity.PlayerInventoryAccess;
import work.lclpnet.kibu.hook.HookRegistrar;
import work.lclpnet.kibu.hook.entity.PlayerInteractionHooks;
import work.lclpnet.kibu.inv.prompt.OptionPrompt;
import work.lclpnet.kibu.scheduler.Ticks;
import work.lclpnet.kibu.translate.text.RootText;
import work.lclpnet.kibu.translate.text.TranslatedText;

import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.UUID;
import java.util.function.Function;

public class KitHandler {

    private static final MapCodec<Boolean> KIT_SELECTOR_CODEC = Codec.BOOL.fieldOf("ap2:kit_selector");
    private static final Item KIT_SELECTOR_ITEM = Items.NETHER_STAR;

    @Getter
    private final KitManager manager;
    private final Participants participants;
    private final KitHandle kitHandle;
    private final Set<UUID> mayChangeKit = new HashSet<>();

    public KitHandler(KitManager manager, Participants participants, KitHandle kitHandle) {
        this.manager = manager;
        this.participants = participants;
        this.kitHandle = kitHandle;
    }

    public void init(HookRegistrar hooks) {
        hooks.registerHook(PlayerInteractionHooks.USE_ITEM, (_player, world, hand) -> {
            if (!(_player instanceof ServerPlayer player) || !participants.isParticipating(player)) {
                return InteractionResult.PASS;
            }

            ItemStack stack = player.getItemInHand(hand);

            if (isKitSelector(stack) && canChangeKit(player)) {
                openKitSelector(player);
                return InteractionResult.SUCCESS_SERVER;
            }

            return InteractionResult.PASS;
        });
    }

    public synchronized void openKitSelector(ServerPlayer player) {
        RootText title = kitHandle.translations().translateText(player, "ap2.kit_selector");
        List<Kit> kits = manager.getKits();

        OptionPrompt.open(player, title, kits, kit -> kitHandle.createKitIcon(kit, player))
                .thenAccept(optKit -> optKit
                        .ifPresent(kit -> changeKit(player, kit)));
    }

    public synchronized void changeKit(ServerPlayer player, Kit kit) {
        if (!mayChangeKit.contains(player.getUUID())) return;

        manager.changeKit(player, kit);

        kitHandle.translations().translateText("ap2.kit_selector.selected", kitHandle.kitName(kit).formatted(ChatFormatting.AQUA))
                .formatted(ChatFormatting.GREEN)
                .sendTo(player);

        player.playNotifySound(SoundEvents.NOTE_BLOCK_PLING.value(), SoundSource.NEUTRAL, 0.5f, 2f);
    }

    public synchronized boolean canChangeKit(ServerPlayer player) {
        return mayChangeKit.contains(player.getUUID());
    }

    public void setupPlayerKits() {
        manager.setupPlayerKits(participants);
    }

    public void enableKitChanger() {
        participants.forEach(this::enableKitChanger);
    }

    public synchronized void enableKitChanger(ServerPlayer player) {
        mayChangeKit.add(player.getUUID());

        var stack = new ItemStack(KIT_SELECTOR_ITEM);

        stack.set(DataComponents.ITEM_NAME, kitHandle.translations().translateText(player, "ap2.kit_selector")
                .formatted(ChatFormatting.AQUA));

        CustomNbt.set(stack, KIT_SELECTOR_CODEC, true);

        player.getInventory().setItem(manager.getOptions().kitSelectorSlot(), stack);
    }

    public void disableKitChanger() {
        participants.forEach(this::disableKitChanger);
    }

    public synchronized void disableKitChanger(ServerPlayer player) {
        mayChangeKit.remove(player.getUUID());

        closeKitChanger(player);

        player.getInventory().removeItemNoUpdate(manager.getOptions().kitSelectorSlot());
    }

    public void closeKitChanger() {
        participants.forEach(this::closeKitChanger);
    }

    public void closeKitChanger(ServerPlayer player) {
        player.closeContainer();
    }

    public boolean isKitSelector(ItemStack stack) {
        if (!stack.is(KIT_SELECTOR_ITEM)) return false;

        return CustomNbt.get(stack, KIT_SELECTOR_CODEC).orElse(false);
    }

    public void selectKitChanger() {
        for (ServerPlayer player : participants) {
            PlayerInventoryAccess.setSelectedSlot(player, manager.getOptions().kitSelectorSlot());
        }
    }

    public void selectKitItem() {
        for (ServerPlayer player : participants) {
            PlayerInventoryAccess.setSelectedSlot(player, manager.getOptions().mainItemSlot());
        }
    }

    public void startKitSelectionTimer(GameCommons commons, Runnable onComplete) {
        startKitSelectionTimer(commons, Ticks.seconds(10), onComplete);
    }

    public void startKitSelectionTimer(GameCommons commons, int ticks, Runnable onComplete) {
        if (manager.getKits().size() < 2) {
            onComplete.run();
            return;
        }

        commons.announcer().announceSubtitle("ap2.kit_selector.hint");

        selectKitChanger();

        TranslatedText label = kitHandle.translations().translateText("ap2.kit_selection");

        commons.createTimerTicks(label, ticks).whenDone(onComplete);
    }

    /**
     * Performs common kit setup.
     * This includes:
     * - initializing the kit manager
     * - initializing all registered kits
     * - enabling the kit changer item, giving the item to the players
     */
    public void setup() {
        manager.init();

        init(kitHandle.hooks());
        setupPlayerKits();
        enableKitChanger();
        selectKitChanger();
    }

    public static KitHandler create(MiniGameHandle gameHandle, ServerLevel world, Function<KitHandle, List<Kit>> kitsFactory) {
        var readView = new ProxyKitReadView();
        var handle = RecordKitHandle.of(gameHandle, world.registryAccess(), readView);

        var manager = new KitManager(kitsFactory.apply(handle));

        readView.inject(manager);

        return new KitHandler(manager, gameHandle.getParticipants(), handle);
    }
}

File: src/lib/src/main/kotlin/work/lclpnet/ap2/game/kit/KitHandler.kt
package work.lclpnet.ap2.game.kit

import com.mojang.serialization.Codec
import com.mojang.serialization.MapCodec
import net.fabricmc.fabric.api.event.player.UseItemCallback
import net.minecraft.ChatFormatting
import net.minecraft.core.component.DataComponents
import net.minecraft.server.level.ServerLevel
import net.minecraft.server.level.ServerPlayer
import net.minecraft.sounds.SoundEvents
import net.minecraft.sounds.SoundSource
import net.minecraft.world.InteractionResult
import net.minecraft.world.entity.player.Player
import net.minecraft.world.item.ItemStack
import net.minecraft.world.item.Items
import work.lclpnet.ap2.ext.mc.isOf
import work.lclpnet.ap2.ext.mc.playNotifySound
import work.lclpnet.ap2.game.MiniGameHandle
import work.lclpnet.ap2.game.MiniGameInstance
import work.lclpnet.ap2.game.base.MapGameInstance
import work.lclpnet.ap2.game.player.Participants
import work.lclpnet.ap2.game.util.Announcer
import work.lclpnet.ap2.game.util.createTimer
import work.lclpnet.kibu.access.entity.PlayerInventoryAccess
import work.lclpnet.kibu.access.misc.CustomNbt
import work.lclpnet.kibu.hook.HookRegistrar
import work.lclpnet.kibu.hook.entity.PlayerInteractionHooks
import work.lclpnet.kibu.inv.prompt.OptionPrompt
import java.util.*
import kotlin.time.Duration
import kotlin.time.Duration.Companion.seconds

private val KIT_SELECTOR_CODEC: MapCodec<Boolean> = Codec.BOOL.fieldOf("ap2:kit_selector")
private val KIT_SELECTOR_ITEM = Items.NETHER_STAR

class KitHandler(
    val manager: KitManager,
    private val participants: Participants,
    private val kitHandle: KitHandle
) {
    private val mayChangeKit: MutableSet<UUID> = HashSet<UUID>()

    fun init(hooks: HookRegistrar) {
        PlayerInteractionHooks.USE_ITEM.registerWith(
            hooks,
            UseItemCallback { player: Player, _, hand ->
                if (player !is ServerPlayer || !participants.isParticipating(player)) {
                    return@UseItemCallback InteractionResult.PASS
                }

                val stack = player.getItemInHand(hand)

                if (isKitSelector(stack) && canChangeKit(player)) {
                    openKitSelector(player)
                    return@UseItemCallback InteractionResult.SUCCESS_SERVER
                }

                InteractionResult.PASS
            })
    }

    @Synchronized
    fun openKitSelector(player: ServerPlayer) {
        val title = kitHandle.translations.translateText(player, "ap2.kit_selector")

        OptionPrompt.open(player, title, manager.kits) {
            kitHandle.createKitIcon(it, player)
        }.thenAccept { optKit ->
            optKit.ifPresent { kit: Kit ->
                changeKit(player, kit)
            }
        }
    }

    @Synchronized
    fun changeKit(player: ServerPlayer, kit: Kit) {
        if (!mayChangeKit.contains(player.getUUID())) return

        manager.changeKit(player, kit)

        kitHandle.translations.translateText(
            "ap2.kit_selector.selected",
            kitHandle.kitName(kit).withStyle(ChatFormatting.AQUA)
        )
            .withStyle(ChatFormatting.GREEN)
            .sendTo(player)

        player.playNotifySound(SoundEvents.NOTE_BLOCK_PLING.value(), SoundSource.NEUTRAL, 0.5f, 2f)
    }

    @Synchronized
    fun canChangeKit(player: ServerPlayer): Boolean =
        mayChangeKit.contains(player.getUUID())

    fun setupPlayerKits() {
        manager.setupPlayerKits(participants)
    }

    fun enableKitChanger() {
        participants.forEach {
            enableKitChanger(it)
        }
    }

    @Synchronized
    fun enableKitChanger(player: ServerPlayer) {
        mayChangeKit.add(player.getUUID())

        val stack = ItemStack(KIT_SELECTOR_ITEM)

        stack.set(
            DataComponents.ITEM_NAME,
            kitHandle.translations.translateText(player, "ap2.kit_selector")
                .withStyle(ChatFormatting.AQUA)
        )

        CustomNbt.set(stack, KIT_SELECTOR_CODEC, true)

        player.inventory.setItem(manager.options.kitSelectorSlot, stack)
    }

    fun disableKitChanger() {
        participants.forEach {
            disableKitChanger(it)
        }
    }

    @Synchronized
    fun disableKitChanger(player: ServerPlayer) {
        mayChangeKit.remove(player.getUUID())

        closeKitChanger(player)

        player.inventory.removeItemNoUpdate(manager.options.kitSelectorSlot)
    }

    fun closeKitChanger() {
        participants.forEach {
            closeKitChanger(it)
        }
    }

    fun closeKitChanger(player: ServerPlayer) {
        player.closeContainer()
    }

    fun isKitSelector(stack: ItemStack): Boolean {
        if (!stack.isOf(KIT_SELECTOR_ITEM)) return false

        return CustomNbt.get(stack, KIT_SELECTOR_CODEC).orElse(false) ?: false
    }

    fun selectKitChanger() {
        for (player in participants) {
            PlayerInventoryAccess.setSelectedSlot(player, manager.options.kitSelectorSlot)
        }
    }

    fun selectKitItem() {
        for (player in participants) {
            PlayerInventoryAccess.setSelectedSlot(player, manager.options.mainItemSlot)
        }
    }

    fun startKitSelectionTimer(gameInstance: MapGameInstance, announcer: Announcer, onComplete: Runnable) {
        startKitSelectionTimer(gameInstance, announcer, 10.seconds, onComplete)
    }

    fun startKitSelectionTimer(gameInstance: MiniGameInstance, announcer: Announcer, duration: Duration, onComplete: Runnable) {
        if (manager.kits.size < 2) {
            onComplete.run()
            return
        }

        announcer.announceSubtitle("ap2.kit_selector.hint")

        selectKitChanger()

        val label = kitHandle.translations.translateText("ap2.kit_selection")

        gameInstance.createTimer(label, duration).whenDone(onComplete)
    }

    /**
     * Performs common kit setup.
     * This includes:
     * - initializing the kit manager
     * - initializing all registered kits
     * - enabling the kit changer item, giving the item to the players
     */
    fun setup() {
        manager.init()

        init(kitHandle.hooks)
        setupPlayerKits()
        enableKitChanger()
        selectKitChanger()
    }

    fun reequip(player: ServerPlayer) {
        manager.getKit(player).unequip(player, manager.options)
        manager.getKit(player).equip(player, manager.options)
    }

    companion object {

        @JvmStatic
        fun create(
            gameHandle: MiniGameHandle,
            world: ServerLevel,
            kitsFactory: (KitHandle) -> List<Kit>
        ): KitHandler {
            val readView = ProxyKitReadView()
            val handle: DefaultKitHandle = DefaultKitHandle.of(gameHandle, world.registryAccess(), readView)

            val manager = KitManager(kitsFactory(handle))

            readView.inject(manager)

            return KitHandler(manager, gameHandle.participants, handle)
        }
    }
}
